### ISSUE TITLE
Crash on Thread Join on Ubuntu 16.04 and 18.04

### ISSUE THREAD

**Issue by cwfitzgerald** [RichieSams/FiberTaskingLib#57]
Alright, so this is a bit of a heisenbug. I put the test code in [this repo](https://github.com/cwfitzgerald/fiber_testing) so that you can test it (should compile on all desktop platforms fine). 

On Ubuntu 16.04 and 18.04 on both gcc-5 and clang-6.0 there is a crash somewhere within `jump_fcontext`. The asm says the segfault is on the first line. `pthread_join` is always on the callstack on the main thread. This manifests as a segmentation fault if you are running it under gdb, but just causes the program to spin indefinitely if run outside of the debugger. 

There's unfortunately not much more information I can provide. You can run the preprocessor on it to remove all the preprocessor magic, and if it helpful I can provide a cpp file with the preprocessesing of the crazy macros already done. However, I highly doubt it's a macro problem, as it is just doing what you would normally do without macros.

If you need any more information, let me know. Thanks for this lovely library

**Comment by RichieSams** [comment]
Thanks for the repory amd the example code. I'll take a look later tonight

**Comment by cwfitzgerald** [comment]
Wonderful thank you! Forgot one of the most important things, the crash only happens on release mode on both gcc/clang. Both MSVC debug and release pass cleanly. ~I have attached a debug symbol'd build with core dump. put it inside of a `build` folder and sources should match.~ I now can't get the damn thing to crash. I did run it with the thread sanitizer however, and I have attached the output there. 

[errors.txt](https://github.com/RichieSams/FiberTaskingLib/files/2346632/errors.txt)

**Comment by cwfitzgerald** [comment]
So you might need to ignore those thread sanitizer warnings, they appear to be false positives due to TSan's handling of atomics. [Reference](https://stackoverflow.com/questions/37552866/why-does-threadsanitizer-report-a-race-with-this-lock-free-example).

I did get it to rash with clang-6.0 on my laptop. I have attached the files below. :)

[core-dump.zip](https://github.com/RichieSams/FiberTaskingLib/files/2346665/core-dump.zip)

**Comment by cwfitzgerald** [comment]
Alright, I manged to get some more information through valgrind. Here are the logs for both gcc and clang.

CMake Call (clang):
```
cmake ../fiber_tasking/ -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-g" -GNinja -DFTL_VALGRIND=On -DCMAKE_ASM_FLAGS="-g" -DCMAKE_CXX_COMPILER=clang++-6.0 -DCMAKE_C_COMPILER=clang-6.0
```

CMake Call (gcc):
```
cmake ../fiber_tasking/ -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-g" -GNinja -DFTL_VALGRIND=On -DCMAKE_ASM_FLAGS="-g" -DCMAKE_CXX_COMPILER=g++-7 -DCMAKE_C_COMPILER=gcc-7
```

Valgrind call (both)
```
valgrind --track-origins=yes ./fiber_tasking 2> valgrind-clang.txt
```

[valgrind-gcc.txt](https://github.com/RichieSams/FiberTaskingLib/files/2350776/valgrind-gcc.txt)
[valgrind-clang.txt](https://github.com/RichieSams/FiberTaskingLib/files/2350777/valgrind-clang.txt)

**Comment by cwfitzgerald** [comment]
Adding `-fno-omit-frame-pointer` seems to help valgrind give a better output. Updated valgrind results below.

[valgrind-gcc.txt](https://github.com/RichieSams/FiberTaskingLib/files/2351132/valgrind-gcc.txt)
[valgrind-clang.txt](https://github.com/RichieSams/FiberTaskingLib/files/2351133/valgrind-clang.txt)

**Comment by RichieSams** [comment]
It looks like a similar issue is still cropping up: https://travis-ci.org/RichieSams/FiberTaskingLib/jobs/432645417

I'm not sure if it's related.

**Comment by cwfitzgerald** [comment]
Alright, it might be. I'll check in when I get home. Though the fact that the later builds passed is a bit weird and might indicate it's an unrelated crash.

**Comment by RichieSams** [comment]
It might be a threading / race bug, that only occurs intermittently on runs.

**Comment by cwfitzgerald** [comment]
TravisCI containers have 2 cores.
It's struck again, this time on g++5 over g++7 https://travis-ci.org/RichieSams/FiberTaskingLib/jobs/432724121

**Comment by RichieSams** [comment]
Optimistically fixed with 89cfa5153fbb5ae28b0f51f39d1fb8ef1fc5e35c
